src/java.base/share/classes/java/time/ZoneOffset.java

```diff
@@ -424,11 +424,17 @@ public static ZoneOffset ofTotalSeconds(int totalSeconds) {
             throw new DateTimeException("Zone offset not in valid range: -18:00 to +18:00");
         }
         if (totalSeconds % (15 * SECONDS_PER_MINUTE) == 0) {
-            return SECONDS_CACHE.computeIfAbsent(totalSeconds, totalSecs -> {
-                ZoneOffset result = new ZoneOffset(totalSecs);
+            Integer totalSecs = totalSeconds;
+            ZoneOffset result = SECONDS_CACHE.get(totalSecs);
```

Comment:
Here, each call may allocate an Integer object. The maximum number of ZoneOffsets that need to be cached here is only 148. Using AtomicReferenceArray is better than AtomicConcurrentHashMap.

Reply:
For example:
```java
    static final class Cache {
              static final AtomicReferenceArray<ZoneOffset> MINUTES_15_CACHE = new AtomicReferenceArray<>(18 * 2 * 4);
    }
    public static ZoneOffset ofTotalSeconds(int totalSeconds) {
        // ...
        int minutes15Rem = totalSeconds / (15 * SECONDS_PER_MINUTE);
        if (totalSeconds - minutes15Rem * 15 * SECONDS_PER_MINUTE == 0) {
            int cacheIndex = minutes15Rem + 18;
            ZoneOffset result = Cache.MINUTES_15_CACHE.get(cacheIndex);
            if (result == null) {
                result = new ZoneOffset(totalSeconds);
                if (!Cache.MINUTES_15_CACHE.compareAndSet(cacheIndex, null, result)) {
                    result = Cache.MINUTES_15_CACHE.get(minutes15Rem);
                }
            }
            return result;
        }
       // ...
    }
```

Reply:
Hi Shaojin,
Thanks for the suggestion, but I am not planning to improve the code more that backing out the offending fix at this time. (btw, cache size would be 149 as 18:00 and -18:00 are inclusinve)

Reply:
Can I submit a PR to make this improvement?

Reply:
@wenshao I agree with your proposal. Also for this part:
```java
ZoneOffset result = MINUTES_15_CACHE.get(cacheIndex);
if (result == null) {
    result = new ZoneOffset(totalSeconds);
    if (!MINUTES_15_CACHE.compareAndSet(cacheIndex, null, result)) {
        result = MINUTES_15_CACHE.get(minutes15Rem);
    }
}
```

I recommend a rewrite:

```java
ZoneOffset result = MINUTES_15_CACHE.getPlain(cacheIndex);
if (result == null) {
    result = new ZoneOffset(totalSeconds);
    ZoneOffset existing = MINUTES_15_CACHE.compareAndExchange(cacheIndex, null, result);
    return existing == null ? result : existing;
}
```

The `getPlain` is safe because if `ZoneOffset` is thread safe, so you can use the object when you can observe its object reference.  Also `compareAndExchange` avoids extra operations if we failed to racily set the computed `ZoneOffset`.